src/geometry/cuda/edge_coarsening_impl.cu

```diff
@@ -0,0 +1,261 @@
+  CUDA_CALL(cudaFree(prop));
+  return done_h;
+}
+
+template<typename FloatType, typename IdType>
+void Propose(const NDArray indptr, const NDArray indices,
+             const NDArray weight, NDArray result, NDArray proposal) {
+  auto* thr_entry = runtime::CUDAThreadEntry::ThreadLocal();
+  const int64_t num_nodes = result->shape[0];
+  IdType *indptr_data = static_cast<IdType*>(indptr->data);
+  IdType *indices_data = static_cast<IdType*>(indices->data);
+  IdType *result_data = static_cast<IdType*>(result->data);
+  IdType *proposal_data = static_cast<IdType*>(proposal->data);
+
+  if (aten::IsNullArray(weight)) {
+    CUDA_KERNEL_CALL(propose_kernel, BLOCKS(num_nodes), THREADS, 0, thr_entry->stream,
+                     indptr_data, indices_data, num_nodes, proposal_data, result_data);
+  } else {
+    FloatType *weight_data = static_cast<FloatType*>(weight->data);
+    CUDA_KERNEL_CALL(weighted_propose_kernel, BLOCKS(num_nodes), THREADS, 0, thr_entry->stream,
+                     indptr_data, indices_data, weight_data, num_nodes, proposal_data, result_data);
+  }
+}
+
+template<typename FloatType, typename IdType>
+void Respond(const NDArray indptr, const NDArray indices,
+             const NDArray weight, NDArray result, NDArray proposal) {
+  auto* thr_entry = runtime::CUDAThreadEntry::ThreadLocal();
+  const int64_t num_nodes = result->shape[0];
+  IdType *indptr_data = static_cast<IdType*>(indptr->data);
+  IdType *indices_data = static_cast<IdType*>(indices->data);
+  IdType *result_data = static_cast<IdType*>(result->data);
+  IdType *proposal_data = static_cast<IdType*>(proposal->data);
+
+  if (aten::IsNullArray(weight)) {
+    CUDA_KERNEL_CALL(respond_kernel, BLOCKS(num_nodes), THREADS, 0, thr_entry->stream,
+                     indptr_data, indices_data, num_nodes, proposal_data, result_data);
+  } else {
+    FloatType *weight_data = static_cast<FloatType*>(weight->data);
+    CUDA_KERNEL_CALL(weighted_respond_kernel, BLOCKS(num_nodes), THREADS, 0, thr_entry->stream,
+                     indptr_data, indices_data, weight_data, num_nodes, proposal_data, result_data);
+  }
+}
+
+template <DLDeviceType XPU, typename FloatType, typename IdType>
+void EdgeCoarsening(const NDArray indptr, const NDArray indices,
+                    const NDArray weight, NDArray result) {
+  // get random generator
+  curandGenerator_t gen;
+  uint64_t seed = dgl::RandomEngine::ThreadLocal()->RandInt(UINT64_MAX);
+  CURAND_CALL(curandCreateGenerator(&gen, CURAND_RNG_PSEUDO_DEFAULT));
+  CURAND_CALL(curandSetPseudoRandomGeneratorSeed(gen, seed));
+
+  // create proposal tensor
+  const int64_t num_nodes = result->shape[0];
+  IdArray proposal = aten::Full(-1, num_nodes, sizeof(IdType) * 8, result->ctx);
+
+  while (!Colorize<IdType>(result, gen)) {
+    Propose<FloatType, IdType>(indptr, indices, weight, result, proposal);
+    Respond<FloatType, IdType>(indptr, indices, weight, result, proposal);
+  }
```

Comment:
Could you please explain why this algorithm is equivalent to the CPU one?

Reply:
This GPU algorithm produces matchings of similar (random matching) or better (weighted matching) quality than CPU version according to the paper. I explain the GPU version in the docstring of `WeightedNeighborMatching` at `src/geometry/cuda/edge_coarsening_impl.cu`